### PR TITLE
store/tikv: mock tikv-server cluster

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: store/tikv/mock-tikv/cluster.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+	"github.com/golang/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/metapb"
+)
+
+// Cluster simulates a TiKV cluster. It focuses on management and the change of
+// meta data. A Cluster mainly includes following 3 kinds of meta data:
+// 1) Region: A Region is a fragment of TiKV's data whose range is [start, end).
+//    The data of a Region is duplicated to multiple Peers and distributed in
+//    multiple Stores.
+// 2) Peer: A Peer is a replica of a Region's data. All peers of a Region forms
```
Comment (shenli) [inline store/tikv/mock-tikv/cluster.go]: forms->form

File: store/tikv/mock-tikv/cluster.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+//    a group, each group elects a Leader to provide services.
+// 3) Store: A Store is a storage/service node. Try to think it as a TiKV server
+//    process. It only responds to client's request if the request's Region's
```
Comment (shenli) [inline store/tikv/mock-tikv/cluster.go]: Only the store with request's Region's leader Peer could respond to client's request.

File: store/tikv/mock-tikv/cluster.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+//    leader Peer is on the Store.
+type Cluster struct {
+	mu      sync.RWMutex
+	id      uint64
+	stores  map[uint64]*Store
+	regions map[uint64]*Region
+}
+
+// NewCluster creates an empty cluster. It needs to be bootstrapped before
+// service.
```
Comment (shenli) [inline store/tikv/mock-tikv/cluster.go]: before providing service

File: store/tikv/mock-tikv/cluster.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+func NewCluster() *Cluster {
+	return &Cluster{
+		stores:  make(map[uint64]*Store),
+		regions: make(map[uint64]*Region),
+	}
+}
+
+// AllocID creates an unique ID in cluster. The ID could be used as either
+// StoreID, RegionID, or PeerID.
```
Comment (shenli) [inline store/tikv/mock-tikv/cluster.go]: StoreID, RegionID or PeerID.

Comment (disksing) [inline store/tikv/mock-tikv/cluster.go]: @shenli This style could help resolving ambiguity. https://en.wikipedia.org/wiki/Serial_comma#Resolving_ambiguity

But in our case I think both is OK :)

File: store/tikv/mock-tikv/cluster.go [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+func (r *Region) updateKeyRange(start, end []byte) {
+	r.meta.StartKey = start
+	r.meta.EndKey = end
+	r.incVersion()
+}
+
+func (r *Region) incConfVer() {
+	r.meta.RegionEpoch = &metapb.RegionEpoch{
```
Comment (shenli) [inline store/tikv/mock-tikv/cluster.go]: Need lock？

Comment (disksing) [inline store/tikv/mock-tikv/cluster.go]: @shenli No. All operations on a Region are through the Cluster, which has a Mutex.

File: store/tikv/mock-tikv/cluster.go [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+func (r *Region) incVersion() {
+	r.meta.RegionEpoch = &metapb.RegionEpoch{
+		ConfVer: r.meta.GetRegionEpoch().ConfVer,
+		Version: proto.Uint64(r.meta.GetRegionEpoch().GetVersion() + 1),
+	}
+}
+
+// Store is the Store's meta data.
+type Store struct {
```
Comment (coocood) [inline store/tikv/mock-tikv/cluster.go]: Why not just use the *metapb.Store directly.

Comment (disksing) [inline store/tikv/mock-tikv/cluster.go]: @coocood We may add methods such as `UpdateAddress()` in the future for test use.

**Comment by coocood** [comment]
LGTM

**Comment by disksing** [comment]
@shenli PTAL.

**Comment by shenli** [comment]
LGTM

**Pull request merged**